File: AllJoynServerExperiences/cs/CryptographicHelpers.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using System;
using System.Text;
using Windows.Security.Cryptography;
using Windows.Storage.Streams;

namespace AllJoynServerExperiences
{
    class CryptographicHelpers
    {
        public static string GenerateHighEntropyKey(int desiredKeyLength)
        {
            if (desiredKeyLength <= 0)
            {
                throw new ArgumentException("The desired key length must be a positive integer.", "desiredKeyLength");
            }
            else
            {
                const int MaxAttempts = 100;
                int attemptCount = 0;
                StringBuilder sb = new StringBuilder(desiredKeyLength);
                IBuffer randomBuffer;
                string str;

                // Base64 uses 6 bits to encode a single character. Number of bits required = length * 6.
                uint numberOfRandomBytesRequired = (uint)(Math.Round((desiredKeyLength * 6.0) / 8.0));

                while ((sb.Length != desiredKeyLength) && (attemptCount < MaxAttempts))
                {
                    randomBuffer = CryptographicBuffer.GenerateRandom(numberOfRandomBytesRequired);
                    str = CryptographicBuffer.EncodeToBase64String(randomBuffer);

                    foreach (char c in str)
                    {
                        if (((c >= '0') && (c <= '9')) || ((c >= 'A') && (c <= 'Z')) || ((c >= 'a') && (c <= 'z')))
                        {
                            sb.Append(c);
                            if (sb.Length == desiredKeyLength)
                            {
                                break;
                            }
                        }
                    }
                    attemptCount++;
                }

                if (sb.Length == desiredKeyLength)
                {
                    return sb.ToString();
                }
                else
                {
                    throw new Exception("Unable to generate a high entropy key.");
                }
            }
        }

        public static string InsertHyphens(string key, int insertHyphenAfterPosition)
        {
            if (string.IsNullOrWhiteSpace(key))
            {
                throw new ArgumentException("Input key must be a valid string.", "key");
            }
            else if (insertHyphenAfterPosition <= 0)
            {
                throw new ArgumentException("Desired hyphen position must be a positive number.", "insertHyphenAfterPosition");
            }
            else if (insertHyphenAfterPosition > key.Length)
            {
                throw new ArgumentException("Desired hyphen position cannot be greater than key length.", "insertHyphenAfterPosition");
            }
            else
            {
                int keyLengthWithHyphens = key.Length + ((key.Length + insertHyphenAfterPosition - 1) / insertHyphenAfterPosition) - 1;
                int count = 0;

                StringBuilder sb = new StringBuilder(keyLengthWithHyphens);
                foreach (char c in key)
                {
                    if (count == insertHyphenAfterPosition)
                    {
                        sb.Append('-');
                        count = 0;
                    }

                    sb.Append(c);
                    count++;
                }

                return sb.ToString();
            }
        }
    }
}

File: Samples/AllJoyn/ProducerExperiences/js/js/cryptographichelpers.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {

    "use strict";

    var cryptoBuffer = Windows.Security.Cryptography.CryptographicBuffer;

    var generateHighEntropyKey = function (desiredKeyLength) {
        if (desiredKeyLength <= 0) {
            throw ("Argument Exception: The desired key length must be a positive integer.");
        } else {
            var maxAttempts = 100;
            var attemptCount = 0;
            var randomBuffer;
            var str;
            var generatedKey = new String();

            // Base64 uses 6 bits to encode a single character. Number of bits required = length * 6.
            var numberOfRandomBytesRequired = Math.round((desiredKeyLength * 6.0) / 8.0);

            while ((generatedKey.length !== desiredKeyLength) && (attemptCount < maxAttempts)) {
                randomBuffer = cryptoBuffer.generateRandom(numberOfRandomBytesRequired);
                str = cryptoBuffer.encodeToBase64String(randomBuffer);

                for (var i = 0; i < str.length; i++) {
                    var c = str.charAt(i);
                    if (/^[a-zA-Z0-9]+$/i.test(c)) {
                        generatedKey += c;
                        if (generatedKey.length === desiredKeyLength) {
                            break;
                        }
                    }
                }
                attemptCount++;
            }

            if (generatedKey.length === desiredKeyLength) {
                return generatedKey;
            } else {
                throw ("Exception: Unable to generate a high entropy key.");
            }
        }
    }

    var insertHyphens = function (key, insertHyphenAfterPosition) {
        if (!key || 0 === key.length) {
            throw ("Argument Exception: Input key must be a valid string.");
        } else if (insertHyphenAfterPosition <= 0) {
            throw ("Argument Exception: Desired hyphen position must be a positive number.");
        } else if (insertHyphenAfterPosition > key.length) {
            throw ("Argument Exception: Desired hyphen position cannot be greater than key length.");
        } else {
            var count = 0;
            var resultString = new String();

            for (var i = 0; i < key.length; i++) {
                if (count === insertHyphenAfterPosition) {
                    resultString += "-";
                    count = 0;
                }

                resultString += key.charAt(i);
                count++;
            }

            return resultString;
        }
    }

    WinJS.Namespace.define("CryptographicHelpers", {
        generateHighEntropyKey: generateHighEntropyKey,
        insertHyphens: insertHyphens
    });
})();